### PR TITLE
Fix dmlc profile

### PR DESCRIPTION
**Pull request by zhang-peng-j**
hotshot is not available in Python 3. So setting DMLC_PROFILE would make dmlc fail with following error message:

_ModuleNotFoundError: No module named 'hotshot'_

This pull request fixes the issue by replacing hotshot with cProfile. A unit test was added to cover DMLC_PROFILE.

### CODE REVIEW

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9719921

File: py/dml/dmlc.py
Comment (zhang-peng-j) [reply py/dml/dmlc.py]: Ok

**Comment by zhang-peng-j** [comment]
> test failure from spellchecker:
> 
> ```
> /disk2/mp/builds/mono-44534/src/tools/dmlc/test/tests.py:193.49-193.59:         'extraenv',                     # Extra envrionment variables
>                                                                                                                         ~~~~~~~~~~~
> 'envrionment' may be a typo of: environment environments
> ```

Fixed. Sorry about the typo.

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9721572

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9721571

File: test/tests.py
Comment (mandolaerik) [inline test/tests.py:741]: Better to keep this code in a separate CTestCase subclass (e.g., you could override test() or run_dmlc())

Comment (zhang-peng-j) [reply test/tests.py]: Ok. I introduced DMLCProfileTestCase, a subclass of CTestCase.

**Comment by zhang-peng-j** [comment]
> You can consider adding a note in README.md, where we have a list of relevant env variables (not required).
> 
> (DMLC_PROFILE is arguably a bit different from most of the current documented variables: existing variables are things you want in .bashrc, while DMLC_PROFILE is something you enable explicitly when you need it)

I documented DMLC_PROFILE anyway. Thanks.

**Review by lwaern-intel — APPROVED** [review]
Please squash together your commits into one and force-push it up, then choose rebase-and-merge on GH.

**Comment by lwaern-intel** [comment]
Sorry, not one: squash together your commits into independent units. Typically, that reduces to one commit, but in this case it might be two, like with with your original two commits you created the PR with.

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9737618

File: test/tests.py
Comment (mandolaerik) [reply test/tests.py:798]: > Please squash together your commits into ~one~ independent units and force-push them up, then choose rebase-and-merge on GH.

I don't think Peng has permissions for merging, so we'll have to push the button after squashing.

Comment (zhang-peng-j) [reply test/tests.py:798]: Ok. I will squash together the commits. But I am now engaged in some urgent cases. So, I will try to do that tomorrow and even later.

BTW,  could you point out what is wrong in the __slots__ definition? I cannot access the test report though I can see that TeamCity build failed....

Comment (lwaern-intel) [reply test/tests.py:798]: Sure:
```
src/tools/dmlc/test/tests.py:798:1: no __slots__ defined in this class [parent-slots]
src/tools/dmlc/test/tests.py:547:1: but parent class CTestCase has __slots__
```
Which can be fixed by:
```
class DMLCProfileTestCase(CTestCase):
    __slots__ = ()
    ...
```

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9741812

File: test/tests.py
Comment (zhang-peng-j) [reply test/tests.py:798]: Thank you! I copied your homework and that saved me much time :-) I squashed together the commits and, indeed, I don't have rebase-and-merge button. Hope I didn't misunderstand anything. Please review. Thank you again.

**Review by lwaern-intel — APPROVED** [review]
I thought PR authors would be able to press the button once the PR was approved, but perhaps not. Thanks for this!

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9741811

**Pull request merged**
